### PR TITLE
Add trivy-k8s scan support (closes #1411)

### LINKED ISSUE #1411: Add trivy k8s scan support
## ➹ New Feature implementation request
Hi,
could we add support for trivy k8s scans, so that users could easily scan their kubernetes config via secureCodeBox?
I think, that it should be easy to trigger such a scan with something like that:
```
apiVersion: execution.securecodebox.io/v1
kind: ScanType
metadata:
  name: trivy-k8s
spec:
  extractResults:
    location: /home/securecodebox/trivy-results.json
    type: trivy-json
  jobTemplate:
    spec:
      backoffLimit: 3
      template:
        spec:
          containers:
            - command:
                - trivy
                - k8s
          serviceAccountName : <some-account-used-to-query-the-k8s-api>
```
```
apiVersion: "execution.securecodebox.io/v1"
kind: Scan
metadata:
  name: "rbac-trivy-k8s"
spec:
  scanType: "trivy-k8s"
  parameters:
    - "--security-checks=rbac"
    - "all"
    - "--report=all"
    - "--format=json"
    - "--output=/home/securecodebox/trivy-results.json"
  ```
However, i am not sure if the parser can handle the ouput.

### PR DESCRIPTION
**Pull request by fbelter-iteratec**
closes https://github.com/secureCodeBox/secureCodeBox/issues/1411

### Checklist

* [ ] Test your changes as thoroughly as possible before you commit them. Preferably, automate your test by unit/integration tests.
* [ ] Make sure that all your commits are signed-off and that you are added to the [Contributors](https://github.com/secureCodeBox/secureCodeBox/blob/main/CONTRIBUTORS.md) file.
* [ ] Make sure that all CI finish successfully.
* [ ] Optional (but appreciated): Make sure that all commits are [Verified](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits).

### CODE REVIEW

File: scanners/trivy/values.yaml [hunk 1]
```diff
@@ -100,6 +100,9 @@ scanner:
   # scanner.tolerations -- Optional tolerations settings that control how the scanner job is scheduled (see: https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/)
   tolerations: []
 
+# kubeauditScope -- Automatically sets up rbac roles for kubeaudit to access the resources it scans. Can be either "cluster" (ClusterRole) or "namespace" (Role)
+kubeauditScope: "cluster"
+
   # -- if set to true the scan job will be suspended after creation. You can then resume the job using `kubectl resume <jobname>` or using a job scheduler like kueue
   suspend: false
```
Comment (Ilyesbdlala) [inline scanners/trivy/values.yaml:108]: I get the following error when trying to deploy the scanner:
```bash
Error: cannot load values.yaml: error converting YAML to JSON: yaml: line 106: did not find expected key
```
You want .Values.kubeauditScope so it should look like the following

```suggestion
  # -- if set to true the scan job will be suspended after creation. You can then resume the job using `kubectl resume <jobname>` or using a job scheduler like kueue
  suspend: false
  
kubeauditScope: "cluster"

```

File: scanners/trivy/parser/parser.js [hunk 2]
```diff
@@ -67,30 +83,123 @@ return findings;
...
-      category = "Go Package Vulnerability";
+function parseK8sScanResults(clusterName, scanResults) {
+  return new Promise((resolve, reject) => {
+
+    var keys = Object.keys(scanResults);
+    const expectedTopLevelAttributes = ["ClusterName", "Vulnerabilities", "Misconfigurations"];
+    const found = keys.find(key => !expectedTopLevelAttributes.includes(key));
+    if (found !== undefined) {
+      reject(new Error("Unexpected attribute '" + found + "' in scan result"));
+    }
```
Comment (Ilyesbdlala) [inline scanners/trivy/parser/parser.js:94]: The parser shows the following error:
```
Parser failed with error:
Error: Unexpected attribute 'Resources' in scan result 
at /home/app/parser-wrapper/parser/parser.js:93:14
at new Promise (<anonymous>) 
at parseK8sScanResults (/home/app/parser-wrapper/parser/parser.js:87:10)
at parse (/home/app/parser-wrapper/parser/parser.js:20:18)                                                                                   at main (/home/app/parser-wrapper/parser-wrapper.js:127:22)                                                                                  at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

for the following results file: https://gist.github.com/Ilyesbdlala/bf015907258e7013615464e98624b172#file-trivy-results-json
I'm guessing adding `Ressources` to `expectedTopLevelAttributes` should fix it. Not sure.

File: scanners/trivy/parser/parser.js [hunk 3]
```diff
@@ -2,22 +2,38 @@
...
+  if (typeof fileContent === "string") {
+    if (fileContent.includes("{") && fileContent.includes("}")) {
+      scanResults = JSON.parse(fileContent)
+    } else {
+      // empty file
+      return [];
+    }
+  }
+
+  if (scanResults.hasOwnProperty('ClusterName')) {
```
Comment (Ilyesbdlala) [inline scanners/trivy/parser/parser.js:18]:
```suggestion
  if (Object.prototype.hasOwnProperty.call(scanResults , 'ClusterName')) {
```
I would avoid calling hasOwnProperty() like so. See https://stackoverflow.com/questions/39282873/object-hasownproperty-yields-the-eslint-no-prototype-builtins-error-how-to

**Review by Ilyesbdlala — CHANGES_REQUESTED** [review]
Good work :) 
There seems to be still issues with the parser and I suggested some minor changes.
Also Nice to have: If you could squash/remove the merge commits and rebase the branch to the newest main.

File: scanners/trivy/parser/parser.js (hunk 2 above)
Comment (fbelter-iteratec) [reply scanners/trivy/parser/parser.js]: Structure of scan result file has changed in the meantime. Had to do some adjustments.

File: scanners/trivy/values.yaml (hunk 1 above)
Comment (fbelter-iteratec) [reply scanners/trivy/values.yaml]: Recommendation adopted

File: scanners/trivy/parser/parser.js (hunk 3 above)
Comment (fbelter-iteratec) [reply scanners/trivy/parser/parser.js]: Recommendation adopted

File: scanners/trivy/examples/k8s/scan.yaml [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+apiVersion: "execution.securecodebox.io/v1"
+kind: Scan
+metadata:
+  name: "trivy-k8s"
+spec:  
+  scanType: "trivy-k8s"
+  parameters:
+    - "cluster"
```
Comment (Ilyesbdlala) [inline scanners/trivy/examples/k8s/scan.yaml:12]: Is there maybe an easy way to also add an integration-test for trivy-k8s scantype ? That would help alot in supporting it :)

**Review by Ilyesbdlala — CHANGES_REQUESTED** [review]

File: scanners/trivy/parser/parser.js [hunk 5]
```diff
@@ -64,7 +82,135 @@ const findings = scanResults.Results.flatMap(({ Target: target, Vulnerabilities
...
+}
+
+function convertTrivyK8sFindingToSCBFinding(trivyK8sFinding, clusterName, namespace, kind, k8sName, target, clazz, type, categoryName) {
+  let reference = trivyK8sFinding.PrimaryURL;
+  if (trivyK8sFinding.VulnerabilityID && trivyK8sFinding.DataSource && trivyK8sFinding.DataSource.URL) {
+    reference = {
+      id: trivyK8sFinding.VulnerabilityID,
+      source: trivyK8sFinding.DataSource.URL
+    }
+  }
```
[the remark this answers is not in the document]
Comment (fbelter-iteratec) [reply scanners/trivy/parser/parser.js]: structure fixed

File: scanners/trivy/parser/parser.test.js [hunk 6]
```diff
@@ -56,8 +56,46 @@ test("should properly parse empty json file", async () => {
...
+    {
+      encoding: "utf8",
+    }
+  );
+  const findings = await parse(jsonContent);
+  await expect(validateParser(findings)).resolves.toBeUndefined();
+  expect(findings).toMatchSnapshot();
+});
+
+
```
[the remark this answers is not in the document]
Comment (fbelter-iteratec) [reply scanners/trivy/parser/parser.test.js]: `JSON.parse()` added

File: scanners/trivy/examples/k8s/scan.yaml (hunk 4 above)
Comment (fbelter-iteratec) [reply scanners/trivy/examples/k8s/scan.yaml:12]: integration test added. only structure of result-document is assured in test.

**Review by Ilyesbdlala — APPROVED** [review]

**Pull request merged**